File: library-app/src/main/java/org/example/library/quote/service/QuoteService.java
package org.example.library.quote.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.example.library.common.exception.NotFoundException;
import org.example.library.library_book.repository.LibraryBookRepository;
import org.example.library.quote.domain.Quote;
import org.example.library.quote.dto.QuoteDto;
import org.example.library.quote.dto.QuoteRequest;
import org.example.library.quote.mapper.QuoteMapper;
import org.example.library.quote.repository.QuoteRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Slf4j
public class QuoteService {

    private final QuoteRepository repository;
    private final QuoteMapper mapper;
    private final LibraryBookRepository libraryBookRepository;

    @Transactional(readOnly = true)
    public List<QuoteDto> getByLibraryBookId(Integer libraryBookId, Integer userId) {
        return repository.findByLibraryBookIdAndLibraryBookUserIdOrderByCreatedAtDesc(libraryBookId, userId)
                .stream()
                .map(mapper::toDto)
                .toList();
    }

    @Transactional
    public QuoteDto create(Integer libraryBookId, QuoteRequest request, Integer userId) {
        var libraryBook = libraryBookRepository.findByIdAndUserId(libraryBookId, userId)
                .orElseThrow(() -> new NotFoundException("error.library_book.not_found"));

        var quote = mapper.toEntity(request);
        quote.setLibraryBook(libraryBook);

        var savedQuote = repository.saveAndFlush(quote);
        log.info("[QUOTE_CREATE] User ID: {}, Library Book ID: {}, Quote ID: {}", userId, libraryBookId, savedQuote.getId());
        return mapper.toDto(savedQuote);
    }

    @Transactional
    public QuoteDto update(Integer quoteId, QuoteRequest request, Integer userId) {
        var quote = repository.findByIdAndLibraryBookUserId(quoteId, userId)
                .orElseThrow(() -> new NotFoundException("error.quote.not_found"));

        mapper.update(quote, request);
        var savedQuote = repository.saveAndFlush(quote);
        log.info("[QUOTE_UPDATE] User ID: {}, Quote ID: {}", userId, quoteId);
        return mapper.toDto(savedQuote);
    }

    @Transactional
    public void delete(Integer quoteId, Integer userId) {
        var quote = repository.findByIdAndLibraryBookUserId(quoteId, userId)
                .orElseThrow(() -> new NotFoundException("error.quote.not_found"));

        repository.delete(quote);
        log.info("[QUOTE_DELETE] User ID: {}, Quote ID: {}", userId, quoteId);
    }

}

File: library-ui/src/app/services/quote.service.ts
import {Injectable} from '@angular/core';
import {ApiService} from './api.service';
import {Observable} from 'rxjs';
import {Quote, QuoteRequest} from '../interfaces/quote';

@Injectable({
  providedIn: 'root'
})
export class QuoteService {

  private readonly API_URL = '/quotes';

  constructor(private api: ApiService) { }

  getByLibraryBookId(libraryBookId: number): Observable<Quote[]> {
    return this.api.get(this.API_URL, { params: { libraryBookId } });
  }

  create(libraryBookId: number, request: QuoteRequest): Observable<Quote> {
    return this.api.post(`${this.API_URL}/${libraryBookId}`, { body: request });
  }

  update(quoteId: number, request: QuoteRequest): Observable<Quote> {
    return this.api.put(`${this.API_URL}/${quoteId}`, { body: request });
  }

  delete(quoteId: number): Observable<void> {
    return this.api.delete(`${this.API_URL}/${quoteId}`, {});
  }

}
